File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.java
import java.util.*;

class Solution {
    boolean solution(String s) {
        if (s.charAt(s.length()-1) == '(') {
            return false;
        }
        
        Queue<Character> stack = new LinkedList<>();
        char[] charS = s.toCharArray();
        for (int i = s.length()-1; i >= 0; i--) {
            if (charS[i] == ')') {
                stack.add(charS[i]);
            } else {
                if (!stack.isEmpty()) {
                    stack.poll();
                } else {
                    return false;
                }
            }
        }
        if (!stack.isEmpty()) {
            return false;
        }

        return true;
    }
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    list_s = list(s)

    if list_s[-1] == '(':
        return False

    stack = []
    
    while list_s:
        par = list_s.pop()
        if par == ')':
            stack.append(par)
        else:
            if stack:
                stack.pop()
            else:
                return False
    if stack:
        return False
    return True